From 62c8558b5373cd603a763e2957f7d7bd86b0d772 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 03 Feb 2022 15:05:30 +0200
Subject: [PATCH] [SW-69581] gaudi3: add memory regions

Due to the complexity of SRAM initialization, for phase1 of testing
spsram memory will be temporary used as storage.
first 4k of the spsram are left for the boot info.
Considering overall limited functionality, at the moment no data
overlapping is expected.
After SRAM memory init is implemented, memory regions should be
placed in the SRAM.

Change-Id: I87578c70958917ed4d58c2ec086abca941466189
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index f02a267..f419d61 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -27,6 +27,7 @@
 #include <boot_if.h>
 #include <uart.h>
 #include <spi_flash.h>
+#include <memory.h>
 
 u32 hl_pll_timestamp_base_get(void)
 {
@@ -268,3 +269,74 @@
 {
 	/* TODO: add cpu_dyn_regs asigments */
 }
+
+/*
+ * due to the complexity of SRAM initialization, for phase1 of testing
+ * spsram memory will be temporary used as storage.
+ * first 4k of the spsram are left for the boot info.
+ * due to overall limited functionality, at the moment no data overlapping
+ * is expected.
+ * TODO: after SRAM memory init is implemented, return the memory regions to
+ * SRAM.
+ */
+
+#define GAUDI3_TMP_MEM		(SP_SRAM_BASE + SZ_4K)
+#define GAUDI3_TMP_MEM_SZ	(HL_SP_SRAM_SIZE - SZ_4K)
+
+static struct mem_desc mem_desc[MEM_NOENT] = {
+	[MEM_GP] = {
+		.base = GAUDI3_TMP_MEM,
+		.size = GAUDI3_TMP_MEM_SZ,
+	},
+	[MEM_LKD_PCI] = {
+		.base = GAUDI3_TMP_MEM,
+		.size = GAUDI3_TMP_MEM_SZ,
+	},
+	[MEM_FWUPD_PCI] = {
+		.base = GAUDI3_TMP_MEM,
+		.size = GAUDI3_TMP_MEM_SZ,
+	},
+	[MEM_I2C] = {
+		.base = GAUDI3_TMP_MEM,
+		.size = GAUDI3_TMP_MEM_SZ,
+	},
+};
+
+int preboot_get_dev_spec_mem_region(enum mem_region region, struct mem_desc *desc)
+{
+	if (region >= MEM_NOENT) {
+		preboot_puts_hex("unknown region = ", region, true);
+		return 1;
+	}
+
+	if (!desc)
+		return 1;
+
+	desc->base = mem_desc[region].base;
+	desc->size = mem_desc[region].size;
+
+	return 0;
+}
+
+bool preboot_is_valid_address_range(u64 addr, int len)
+{
+	if (addr >= SP_SRAM_BASE &&
+	    addr + len <= SP_SRAM_BASE + HL_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_SP_SRAM_BASE &&
+	    addr + len <= HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= ARC0_ICCM_FULL_BASE &&
+	    addr + len <= ARC0_ICCM_FULL_BASE + ARC0_MEM_SIZE)
+		return true;
+
+	if (addr >= ARC1_ICCM_FULL_BASE &&
+	    addr + len <= ARC1_ICCM_FULL_BASE + ARC1_MEM_SIZE)
+		return true;
+
+	preboot_puts_hex64("ERROR: invalid address range addr = ", addr, true);
+
+	return false;
+}
